site-src/geps/gep-726.md

```diff
@@ -0,0 +1,266 @@
+# GEP-726: Add Path Redirects and Rewrites
+
+* Issue: [#726](https://github.com/kubernetes-sigs/gateway-api/issues/726)
+* Status: Implementable
+
+## TLDR
+
+This GEP proposes adding support for path redirects and rewrites in addition to
+host rewrites. This would augment the existing host redirection capabilities.
+
+## Goals
+
+* Implement path redirects.
+* Implement the most portable and simple forms of path rewrites.
+* Describe how more advanced rewrite and redirect and redirect capabilities
+  could be added in the future.
+
+## API
+
+Although many implementations support very advanced rewrite and redirect
+capabilities, the following are the most [portable](#portability) concepts that
+are not already supported by the Gateway API:
+
+* Path redirects
+* Path prefix redirects
+* Path prefix rewrites
+* Host rewrites
+
+Although regular expression based redirects and rewrites are commonly supported,
+there is significantly more variation in both if and how they are implemented.
+Given the wide support for this concept, it is important to design the API in a
+way that would make it easy to add this capability in the future.
+
+### Redirects
+
+The existing `RequestRedirect` filter can be expanded to support path redirects:
+
+```yaml
+kind: HTTPRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: http-filter-1
+spec:
+  rules:
+    - matches:
+      - path:
+          type: Prefix
+          value: /
+      filters:
+      - type: RequestRedirect
+        requestRedirect:
+          hostname: foo.com
+          path:
+            type: Prefix
+            value: /bar
+```
+
+This would be represented with the following API additions:
+```go
+// HTTPRequestRedirect defines a filter that redirects a request. At most one of
+// these filters may be used on a Route rule. This may not be used on the same
+// Route rule as a HTTPRequestRewrite filter.
+type HTTPRequestRedirect struct {
+  // Path defines a path redirect.
+  //
+  // Support: Extended
+  //
+  // +optional
+  Path *HTTPPathRedirect `json:"path,omitempty"`
+  // ...
+}
+
+// HTTPPathRedirectType defines the type of path redirect.
+type HTTPPathRedirectType string
+
+const (
+  // This type of redirect indicates that the complete path will be replaced by
+  // the path redirect value.
+  AbsoluteHTTPPathRedirect HTTPPathRedirectType = "Absolute"
```

Comment:
I'm wondering whether these options are unnecessarily constrained? There's really 2 cases

1. Emit the entire location path
2. Generate the location path by replacing the matched portion

In case (2), the matched portion could be a prefix, or a regex, or implementation-defined.

Reply:
On 2 I agree that there's room for other kinds of replacement, but I don't think those are quite as portable. I wanted to start with prefix because that seems to be the most universally supportable concept.

Reply:
So, we are open to adding an ImplementationSpecific option here in the future?

Reply:
I think 2 is actually solved now, and 1 always was. This comment came from a time when the options were "Absolute" and "Prefix" which did not scale with other match types. Given our removal of ImplementationSpecific types elsewhere, I don't think it would make sense to add it back without a clear rationale.